src/pybind/mgr/deepsea/module.py

```diff
@@ -307,13 +318,42 @@ def _set_last_failure_msg(self, msg):
     # Note: this is pretty braindead and doesn't implement the full eventsource
     # spec, but it *does* implement enough for us to listen to events from salt
     # and potentially do something with them.
-    # TODO: How are we going to deal with salt-api dying, or mgr failing over,
-    #       or other unforeseen glitches when we're waiting for particular jobs
-    #       to complete?  What's to stop things falling through the cracks?
-    #       Do completions we're waiting on need to be persisted somewhere?
     def _read_sse(self):
         event = {}
         try:
+            # Just starting the event reader; if we've made it here, we know we're
+            # talking to salt-api (_do_request would have raised an exception if the
+            # response wasn't ok), so check if there's any completions inflight that
+            # need to be dealt with.  This handles the case where some command was
+            # invoked, then salt-api died somehow, and we reconneced, but missed the
+            # completion at the time it actually happened.
+            for tag in list(self._all_completions):
+                self.log.info("Found event {} inflight".format(tag))
+                resp = self._do_request_with_login("POST", data = {
+                    "client": "runner",
+                    "fun": "jobs.lookup_jid",
+                    "jid": tag.split('/')[2]
+                })
+                if resp.ok:
+                    # jobs.lookup_jid returns a dict keyed by hostname.
+                    return_dict = resp.json()['return'][0]
+                    if return_dict:
+                        # If the job is complete, there'll be one item in the dict.
+                        self.log.info("Event {} complete".format(tag))
+                        # The key is the salt master hostname, but we don't care
+                        # about that, so just grab the data.
+                        data = next(iter(return_dict.items()))[1]
```

Comment:
`return_dict` always contains only one entry?

Reply:
Yep (at least, if the thing is complete). The entry is the host on which the runner executed.